src/pybind/mgr/cephadm/ssh.py

```diff
@@ -223,6 +219,10 @@ async def _write_remote_file(self,
                 f.flush()
                 conn = await self._remote_connection(host, addr)
                 await asyncssh.scp(f.name, (conn, tmp_path))
+            if uid is not None and gid is not None and mode is not None:
```

Comment:
do we want this to be an `elif` right under the `if self.mgr.ssh_user != 'root'` block (basically just reversing the order of the two blocks instead of moving it below the `with NamedTemporaryFile(prefix='cephadm-write-remote-file-') as f:` block). I'm worried that, if both `self.mgr.ssh_user != 'root'` and `uid is not None and gid is not None and mode is not None` are true we'll end up just setting it to the given uid/gid/mode and will basically follow the old behavior. For reference, the ceph.conf file, which is causing the issue, is given default uid/gid of 0/0 and that uid/gid are always provided to calls to this function when writing the ceph.conf. So if we want to enforce matching the ssh user, I think we need it to take priority over that which means we have to only use the uid/gid if the ssh user is the default (root). Either that, or you'd have to go change what we generate for uid/gid for ceph.conf https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/serve.py#L1001

Reply:
I thought this would be the expected behavior. The problem is not `ceph.conf`, because the move code below is executed with sudo anyways. The problem is the temporary `ceph.conf.new`. This is not writable in the with block for the non-root-user.
So after writing the `ceph.conf.new` file in the with block, the permissions of that file should be restored to the provided config. So the move of `/tmp/etc/ceph/ceph.conf.new` results in a `/etc/ceph/ceph.conf` with proper permissions.

In essence `/etc/ceph/ceph.conf` remains as it is, only for a short period of time ssh_user is the owner `/tmp/etc/ceph/ceph.conf.new`.

I would not change permissions on /etc/ceph/ceph.conf, because I just don't know what could possibly go wrong on that end ;) At least it could become irritating when seeing only the uid when being in a container.

Reply:
ah, okay. I see what you mean about the temp file being the one that matters rather than the actual ceph.conf. I think this looks good in that case. Thanks again, I'll try to get this through the necessary qa and merge (or let you know if we run into issues).

Reply:
while this is true of the current implementation, I think the original intent was to have the uid/gid set on the resulting file.

please see this from the v16 (pacific) branch:
https://github.com/ceph/ceph/blob/f5e8c20edf1fd6ce9fe4d0f2c8790aa6d1433460/src/pybind/mgr/cephadm/serve.py#L1391

Reply:
so, should we be setting the uid/gid after the mv command?